### PR TITLE
Feat 1350 bridge tx

### PR DESCRIPTION
**Pull request by mhagel**
## Description

Closes: regen-network/regen-registry#1350

- Wires up MsgBridge tx

Screen Shot 2022-11-04 at 1 56 13 PM

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] provided a link to the relevant issue or specification
- [x] provided instructions on how to test
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed
- [ ] once the PR is closed, set up backport PRs for `redwood`, `hambach` and optionally `master` (see below)

#### Setting up backport PRs

After merging your PR to `dev`, set up backports by doing the following:

1. If your branch does not have merge commits, add the following comment to
   your PR:

- If PR is a new feature: `@Mergifyio backport redwood hambach`.
- If PR is a bug fix: `@Mergifyio backport redwood hambach master`.

2. If your branch does have merge commits:

   a. Pull latest `dev`, `hambach` and `redwood` branches

   b. Create new branches for backports and merge `dev` (replace `<PR#>` with your PR #)

   ```
   git checkout -b hambach-backport-<PR#> hambach
   git merge dev
   git push origin hambach-backport-<PR#>
   git checkout -b redwood-backport-<PR#> redwood
   git merge dev
   git push origin redwood-backport-<PR#>`
   ```

   c. Open new PRs in regen-web targeting `hambach` and `redwood`, respectively.

   d. If PR is a bug fix repeat the same process for `master` branch.

### How to test

NOTE: `regen1df675r9vnf7pdedn4sf26svdsem3ugavgxmy46` has some ecocredits with batchDenom starting w/ C46.

2nd NOTE: This PR hard-codes the filter for "Bridgable" credits as `C03` (true for mainnet), but for redwood it is [`C46`](https://redwood.regen.aneka.io/txs/250F791598E6434E357089709826009019E9892292D925BB55BF05A98D62C48A) .

1. Make sure your account (`regen1df675r9vnf7pdedn4sf26svdsem3ugavgxmy46`) has some "Bridgable" ecocredits (aka "originated from polygon")
2. https://deploy-preview-1543--regen-registry.netlify.app/ecocredits/dashboard 
3. click Bridge tab
4. click Bridge button on a bridgable ecocredit
5. go through the Bridge flow (**set amount to 0.001** or any small amount until we have more to play with) and submit the TX
6. Expect success modal
7. Expect **0.001** credits to be subtracted from your "Bridgable" ecocredits
8. Expect a "BRIDGE" tx in /ecocredits/activity

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

File: web-components/src/components/modal/modal.stories.tsx [hunk 1]
```diff
@@ -323,13 +322,3 @@ export const createSellOrderModal = (): JSX.Element => (
     onSubmit={async () => alert('submit')}
   />
 );
-
```
Comment (mhagel) [inline web-components/src/components/modal/modal.stories.tsx:326]: moved to web-registry, so no more need for story

File: web-registry/src/components/organisms/BridgableEcocreditsTable/BridgableEcocreditsTable.tsx [hunk 2]
```diff
@@ -76,96 +85,107 @@ export const BridgableEcocreditsTable = ({
   }
 
   return (
-    <WithLoader isLoading={isLoadingCredits} sx={loaderStyles.withLoaderBlock}>
```
Comment (mhagel) [inline web-registry/src/components/organisms/BridgableEcocreditsTable/BridgableEcocreditsTable.tsx:79]: Just wrapping this in a fragment to include `<BridgeFlow>`.
Only real change starts line 184.

File: web-registry/src/features/marketplace/BridgeFlow/hooks/useCreditBridgeSubmit.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  creditBridgeBatch,
+  signAndBroadcast,
+  setCreditBridgeOpen,
+  setCardItems,
+  setTxModalHeader,
+  setTxModalTitle,
+  setTxButtonTitle,
+  setTxModalDescription,
+}: Props): ReturnType => {
+  const track = useTrack();
```
Comment (mhagel) [inline web-registry/src/features/marketplace/BridgeFlow/hooks/useCreditBridgeSubmit.tsx:45]: I went ahead and added the tracking events, but we can add them separately if this is out of scope. cc @wgwz

File: web-registry/src/components/organisms/BridgableEcocreditsTable/BridgableEcocreditsTable.tsx [hunk 4]
```diff
@@ -76,96 +85,107 @@ export const BridgableEcocreditsTable = ({
...
+      >
+        <ActionsTable
+          tableLabel="bridgable ecocredits table"
+          sx={tableStyles.rootOnlyTopBorder}
+          renderActionButtons={(i: number) => (
+            <OutlinedButton
+              startIcon={<BridgeIcon sx={{ width: '24px', height: '24px' }} />}
+              size="small"
+              onClick={async () => {
+                setBatchToBridge(bridgableCredits[i]);
```
Comment (mhagel) [inline web-registry/src/components/organisms/BridgableEcocreditsTable/BridgableEcocreditsTable.tsx:101]: change here

File: web-registry/src/lib/ecocredit/constants.ts [hunk 5]
```diff
@@ -138,6 +139,10 @@ export const ECOCREDIT_MESSAGE_TYPES = {
     message: `/${MsgBridge.$type}`,
     readable: 'bridge',
   },
+  BRIDGE_RECEIVE: {
```
Comment (mhagel) [inline web-registry/src/lib/ecocredit/constants.ts:142]: We won't be testing this as part of this PR, but I am adding `MsgBridgeReceive` to the activity table here before I forget.

Comment (mhagel) [reply web-registry/src/lib/ecocredit/constants.ts:142]: `MsgBridge` is already in here, so you will see that in /activity if a TX succeeds.

File: web-registry/src/features/marketplace/BridgeFlow/hooks/useCreditBridgeSubmit.tsx [hunk 6]
```diff
@@ -0,0 +1,127 @@
+import { useCallback } from 'react';
```
Comment (mhagel) [inline web-registry/src/features/marketplace/BridgeFlow/hooks/useCreditBridgeSubmit.tsx:1]: Mostly following established patterns here, but worth reviewing.

File: web-registry/src/hooks/bridge/useBridgable.ts [hunk 7]
```diff
@@ -4,33 +4,35 @@ import { TablePaginationParams } from 'web-components/lib/components/table/Actio
...
-      setCredits([]);
-      setIsLoadingCredits(false);
-    });
-  }, []);
-
-  return { credits, isLoadingCredits };
+    if (isLoadingCredits) return;
+    const _bridgableCredits = credits.filter(
+      credit =>
+        credit.classId === 'C03' && Number(credit.balance?.tradableAmount) > 0,
```
Comment (mhagel) [inline web-registry/src/hooks/bridge/useBridgable.ts:32]: `C03` in mainnet, but the Toucan credit class will probably be different in redwood. To test, we will have to temporarliy alter this for local and deploy-preview testing. 

Maybe we should set an env variable for this, so it works correctly in redwood?

Comment (mhagel) [reply web-registry/src/hooks/bridge/useBridgable.ts:32]: I think this will be C46 in Redwood: 
https://redwood--regen-registry.netlify.app/credit-classes/C46

**Comment by clevinson** [comment]
For anyone else wanting to test out this functionality, its important to understand that credits can only be bridged from Regen => Polygon if they originated on Polygon as TCO2s and were previously bridged to Regen (therefore the bridging from Regen to Polygon essentially acts as "bridging back to origination chain").

I've added some basic instructions for getting bridgeable ecocredits here: https://github.com/rnd-inc/toucan-bridge-service/pull/246

File: web-registry/src/hooks/bridge/useBridgable.ts (hunk 7 above)
Comment (mhagel) [reply web-registry/src/hooks/bridge/useBridgable.ts]: Using an env var now, so deploy preview and redwood should work

**Comment by mhagel** [comment]
@erikalogie ready for feedback. I have asked @victordelval to help finish this up. Testing instructions above.

https://deploy-preview-1543--regen-registry.netlify.app/ecocredits/dashboard

**Comment by flagrede** [comment]
> For anyone else wanting to test out this functionality, its important to understand that credits can only be bridged from Regen => Polygon if they originated on Polygon as TCO2s and were previously bridged to Regen (therefore the bridging from Regen to Polygon essentially acts as "bridging back to origination chain").
> 
> I've added some basic instructions for getting bridgeable ecocredits here: https://github.com/rnd-inc/toucan-bridge-service/pull/246

Thanks for the clarification @clevinson. Could you add me to the repo you linked? I don't have access currently.

**Review by flagrede — COMMENTED** [review]
Looks good so far, will give it a second look when it will be finished.
I just noticed that some fields (project, credit class, and location) never load in the table.

File: web-components/src/components/form/BridgeForm.tsx [hunk 8]
```diff
@@ -66,7 +70,7 @@ const BridgeForm = ({
       {({ values, submitForm, isSubmitting, isValid, submitCount, status }) => (
         <Form>
           <Field
-            name="chain"
+            name="target"
             label="Chain"
             component={SelectTextField}
             options={[{ label: 'Polygon', value: 'polygon' }]}
```
Comment (victordelval) [inline web-components/src/components/form/BridgeForm.tsx:76]: @flagrede I'm extracting a constant with the allowed chains for bridge. Currently only one item for Polygon, and the initial values of the form read from this first option

File: web-components/src/components/form/BridgeForm.tsx [hunk 9]
```diff
@@ -41,10 +42,13 @@ const BridgeForm = ({
     recipient: '',
     agreeErpa: false,
     batchDenom,
+    target: 'polygon', //TODO: temp?
```
[the remark this answers is not in the document]
Comment (victordelval) [reply web-components/src/components/form/BridgeForm.tsx:45]: see below

File: web-components/src/components/modal/modal.stories.tsx (hunk 1 above)
Comment (victordelval) [reply web-components/src/components/modal/modal.stories.tsx:326]: Done!

**Comment by victordelval** [comment]
> For anyone else wanting to test out this functionality, its important to understand that credits can only be bridged from Regen => Polygon if they originated on Polygon as TCO2s and were previously bridged to Regen (therefore the bridging from Regen to Polygon essentially acts as "bridging back to origination chain").
> 
> I've added some basic instructions for getting bridgeable ecocredits here: https://github.com/rnd-inc/toucan-bridge-service/pull/246

Hey @clevinson! if possible give me access to read in that repo

**Review by victordelval — APPROVED** [review]
Good work @mhagel!

**Comment by victordelval** [comment]
> @victordelval I guess since you are taking over this task, I'll direct my feedback to you. Is viewing the pending bridged credits part of a separate task? Also, should I be able to see the credits in my Metamask wallet somehow, and if so, how?

Hi @erikalogie! the queries corresponding to the "bridged" table, along with the status column included there, will be completed in the issue [#regen-registry/1351](https://app.zenhub.com/workspaces/regen-registry-5f8998bec8958d000f4609e2/issues/regen-network/regen-registry/1351). As for seeing the credits you have in the polygon bridge with Metamask, since it is a testnet you have to add it manually in Metamask, but I don't have the details...

File: web-components/src/components/form/BridgeForm.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (victordelval) [reply web-components/src/components/form/BridgeForm.tsx]: A clarification. I have kept the BRIDGE_ALLOWED_CHAINS constant on the form itself temporarily while waiting to move the forms to web-registry. In case the forms are not moved from web-components, ideally the allowed chains information should reside in the global web-registry and be passed to the (imported) form.

**Comment by erikalogie** [comment]
> > @victordelval I guess since you are taking over this task, I'll direct my feedback to you. Is viewing the pending bridged credits part of a separate task? Also, should I be able to see the credits in my Metamask wallet somehow, and if so, how?
> 
> Hi @erikalogie! the queries corresponding to the bridged table, along with the status column included there, will be completed in the issue [#regen-registry/1351](https://app.zenhub.com/workspaces/regen-registry-5f8998bec8958d000f4609e2/issues/regen-network/regen-registry/1351). As for seeing the credits you have in the polygon bridge with Metamask, since it is a testnet you have to add it manually in Metamask, but I don't have the details...

@victordelval cool, got it

@clevinson do you know how we can test this so that we can see the ecocredits that are bridged to polygon on some kind of test net in metamask?

**Comment by blushi** [comment]
> @clevinson do you know how we can test this so that we can see the ecocredits that are bridged to polygon on some kind of test net in metamask?

you need to add mumbai testnet to metamask (go to https://mumbai.polygonscan.com/ and click on "add mumbai network" in the bottom right corner), then once connected, open metamask and import the token https://mumbai.polygonscan.com/token/0xf7e61e0084287890e35e46dc7e077d7e5870ae27 from there and you should be able to see your balance
From this token page on polyscan, you should also see the corresponding tx in the transfers table.
You can also go to https://mumbai.polygonscan.com/token/0xf7e61e0084287890e35e46dc7e077d7e5870ae27?a={yourEthAddress} to see your token balance

until we implement regen-network/regen-registry#1351, you can go to https://toucan-bridge-staging.herokuapp.com/regen/events/{regenTxHash}/status (`regenTxHash` is displayed on the confirmation popup) to see what's the status on polygon:
- regen_ready: the event triggering the bridge has been polled and stored in the database.
- evm_broadcast: the event has been processed and the transaction minting the credits on polygon has been sent.
- evm_confirmed: the event has been successfully bridged and the transaction was confirmed.
- error: an error has occurred during the processing of the event. This is an unrecoverable error and the support team should be contacted.

**Comment by blushi** [comment]
Any idea why the project and credit class are not displayed on the bridgable ecocredits table? @victordelval @mhagel 

image

**Comment by victordelval** [comment]
> Any idea why the project and credit class are not displayed on the bridgable ecocredits table? @victordelval @mhagel
> 

@blushi what I have discovered is that in redwood, both the `C46-001` project and the only credit batch that contains `C46-001-20140701-20141231-001`, for the Bridge, both entities have the same IRI for metadata `regen:13toVfvC2YxrrfSXWB5h2BGHiXZURsKxWUz72uDRDSPMCrYPguGUXSC.rdf`, when each one should have their own IRI..

```
$ regen query --node http://redwood.regen.network:26657/ --chain-id regen-redwood-1 ecocredit project C46-001
project:
  admin: regen13mkv806hwwtmr4w7rwu26rm70a3dxhzmcdyll8
  class_id: C46
  id: C46-001
  jurisdiction: ID
  metadata: regen:13toVfvC2YxrrfSXWB5h2BGHiXZURsKxWUz72uDRDSPMCrYPguGUXSC.rdf
  reference_id: VCS-674

$ regen query --node http://redwood.regen.network:26657/ --chain-id regen-redwood-1 ecocredit batches-by-project C46-001
batches:
- denom: C46-001-20140701-20141231-001
  end_date: "2014-12-31T00:00:00Z"
  issuance_date: "2022-11-03T21:45:10.048802092Z"
  issuer: regen13mkv806hwwtmr4w7rwu26rm70a3dxhzmcdyll8
  metadata: regen:13toVfvC2YxrrfSXWB5h2BGHiXZURsKxWUz72uDRDSPMCrYPguGUXSC.rdf
  open: true
  project_id: C46-001
  start_date: "2014-07-01T00:00:00Z"
pagination:
  next_key: null
  total: "1"

```

**Comment by mhagel** [comment]
> Any idea why the project and credit class are not displayed on the bridgable ecocredits table? @victordelval @mhagel

@blushi 

The filter was being applied in the wrong place. Now we are optionally filtering by creditClassId in the paginated `useEcocredit` requests.

Screen Shot 2022-11-14 at 10 34 39 PM

The project name now properly falls back to projectId, and I added C46 to sanity-staging as "Toucan Credits."
When project metadata is available, the project name will pull from that as a first choice.

cc @erikalogie

File: netlify.toml [hunk 10]
```diff
@@ -10,6 +10,7 @@
   REACT_APP_LEDGER_CHAIN_ID = "regen-1"
   REACT_APP_SENTRY_ENVIRONMENT="production"
   REACT_APP_ANALYTICS_ENABLED="true"
+  REACT_APP_BRIDGE_CREDIT_CLASS_ID = "C03"
```
[the remark this answers is not in the document]
Comment (mhagel) [reply netlify.toml:13]: It is possible maybe, but I don't think there are any plans for this.

File: web-registry/src/hooks/useEcocredits.ts [hunk 11]
```diff
@@ -23,12 +24,26 @@ const hasBatchBalance = (batchWithBalance: BatchInfoWithBalance): boolean => {
   );
 };
 
+/* If a creditClassId is provided, filter by that. */
```
Comment (mhagel) [inline web-registry/src/hooks/useEcocredits.ts:27]: To get the pagination right, I had to apply an optional creditClass filter both during the "initial credits" load (line 78), and after the metadata and CMS content are loaded (line 119).

@victordelval @flagrede Let me know if this sounds right to you.

**Comment by blushi** [comment]
> The project name now properly falls back to projectId, and I added C46 to sanity-staging as "Toucan Credits." When project metadata is available, the project name will pull from that as a first choice.

@mhagel do we also just display the credit class id in case we're unable to find the metadata hence the credit class name?

File: web-registry/src/hooks/useEcocredits.ts (hunk 11 above)
Comment (victordelval) [reply web-registry/src/hooks/useEcocredits.ts:27]: LGTM @mhagel!

Comment (victordelval) [reply web-registry/src/hooks/useEcocredits.ts:27]: just a small suggestion for improvement. The predicate function `isOfCreditClass` used to filter, we can curry it and make its use a little more compact and elegant, without the need to pass the items (balances) that are being iterated...

```
const isOfCreditClass =
  (creditClassId?: string) =>
  (balance: BatchBalanceInfo): boolean => {
    if (!creditClassId) return true;
    return balance.batchDenom.startsWith(`${creditClassId}-`);
  };
```
so we can do...
```
...balances.filter(isOfCreditClass(creditClassId))...
```

Comment (mhagel) [reply web-registry/src/hooks/useEcocredits.ts:27]: Yeah I like that. Done.

**Comment by mhagel** [comment]
> @mhagel do we also just display the credit class id in case we're unable to find the metadata hence the credit class name?

@blushi 
This cell was rendering Sanity content.

I just updated `BlockContent.tsx` to make this easier. Now, if we feed a string into `<BlockContent content="C46" />` it will render the string. Before, it just appeared blank if the `content` was not a Sanity block.

And yes, now falling back to credit class ID if no CMS creditClassName.

File: web-components/src/components/block-content/BlockContent.tsx [hunk 12]
```diff
@@ -68,11 +68,15 @@ const CustomBlockContent: React.FC<{
           ...(Array.isArray(sx) ? sx : [sx]),
         ]}
       >
-        <BlockContent
-          className={className}
-          blocks={content}
-          serializers={serializers}
-        />
+        {typeof content === 'string' ? (
```
[the remark this answers is not in the document]
Comment (victordelval) [reply web-components/src/components/block-content/BlockContent.tsx:71]: @mhagel yes, I agree that the complementary information of the on-chain entities should only come from the off-chain source

[the remark this answers is not in the document]
Comment (mhagel) [reply web-components/src/components/block-content/BlockContent.tsx:71]: > yeah I was actually wondering whether we should just not use sanity for pulling the credit class name and just use the on chain credit class metadata instead...

@blushi 
My only problem with this is that we have to make _another_ ledger query for each table result. But this is not a big deal if there is some sort of caching available.

File: web-registry/src/features/marketplace/BridgeFlow/BridgeFlow.tsx [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        onSubmit={creditBridgeSubmit}
+        onClose={() => setBatchToBridge(undefined)}
+        batch={selectedBatch}
+      />
+      <ProcessingModal
+        open={isProcessingModalOpen}
+        onClose={() => setIsProcessingModalOpen(false)}
+      />
+      <TxSuccessfulModal
+        open={!!txHash && !error} //TODO
```
[the remark this answers is not in the document]
Comment (mhagel) [reply web-registry/src/features/marketplace/BridgeFlow/BridgeFlow.tsx]: removed, thx

**Comment by blushi** [comment]
Looks good, just one thing: on the confirmation popup, the "view pending bridged credits" should open up the "bridged ecocredits" subtab

**Comment by mhagel** [comment]
> Looks good, just one thing: on the confirmation popup, the "view pending bridged credits" should open up the "bridged ecocredits" subtab

@blushi  Done. I decided to do this with location.state instead of creating a sub-route. The latter was creating unwanted re-renders with nested tabs.

File: web-components/src/components/block-content/BlockContent.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (flagrede) [reply web-components/src/components/block-content/BlockContent.tsx:73]: Getting credit class names only from metadata sounds better indeed but should probably be handled in a separate PR as it involves several pages and it also requires some work on the caching side as Mark said to make it efficient.

File: web-registry/src/features/marketplace/BridgeFlow/BridgeFlow.tsx [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+import { BRIDGE_HEADER } from './BridgeFlow.constants';
+import useCreditBridgeSubmit from './hooks/useCreditBridgeSubmit';
+
+type Props = {
+  isFlowStarted: boolean;
+  setBatchToBridge: UseStateSetter<BatchInfoWithBalance | undefined>;
+  selectedBatch: BatchInfoWithBalance | undefined;
+};
+
+export const BridgeFlow = ({
```
Comment (flagrede) [inline web-registry/src/features/marketplace/BridgeFlow/BridgeFlow.tsx:31]: Our flows have a lot of code in common so far. Starting to think that we could reduce the boilerplate by extracting modals into the layout.
Nothing to change here but since we'll probably write more flow like this in the future it might be worth investigating.

File: web-registry/src/pages/Dashboard/MyBridge/MyBridge.tsx [hunk 15]
```diff
@@ -36,10 +39,30 @@ export const MyBridge = (): JSX.Element => {
     [wallet?.address],
   );
 
+  useEffect(() => {
+    const _activeTab = state?.tab
+      ? Math.max(
+          tabs.findIndex(tab => tab.label.toLowerCase().includes(state?.tab)),
+          0,
+        )
```
Comment (flagrede) [inline web-registry/src/pages/Dashboard/MyBridge/MyBridge.tsx:47]: Should this be a small utils function?

**Review by flagrede — APPROVED** [review]
Looks good nice work !
Left some minor comments and ideas for later.

File: web-components/src/components/block-content/BlockContent.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (blushi) [reply web-components/src/components/block-content/BlockContent.tsx:73]: agreed, I started a thread on slack about that: https://regen-network.slack.com/archives/C01LX9E8QN8/p1668618332989469

**Review by blushi — APPROVED** [review]

**Comment by mhagel** [comment]
@blushi @flagrede 

Added a new issue to extract CMS dependency from this table: regen-network/regen-registry#1428

File: web-registry/src/pages/Dashboard/MyBridge/MyBridge.tsx (hunk 15 above)
Comment (mhagel) [reply web-registry/src/pages/Dashboard/MyBridge/MyBridge.tsx]: extracted to a hook

File: web-components/src/components/block-content/BlockContent.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (mhagel) [reply web-components/src/components/block-content/BlockContent.tsx:73]: @blushi @flagrede

Added a new issue to extract CMS dependency from this table: https://github.com/regen-network/regen-registry/issues/1428

**Comment by mhagel** [comment]
@erikalogie Ready for another look: https://deploy-preview-1543--regen-registry.netlify.app/ecocredits/bridge

**Comment by mhagel** [comment]
Merging - I believe all concerns have been addressed

**Comment by mhagel** [comment]
@mergifyio backport redwood hambach

**Comment by erikalogie** [comment]
Lgtm! @mhagel
